apps/test/unit/templates/sectionProgress/progressTables/ProgressTableDetailCellTest.jsx

```diff
@@ -0,0 +1,143 @@
+    const wrapper = shallow(<ProgressTableDetailCell {...DEFAULT_PROPS} />);
+    const levelBubble1 = wrapper.findWhere(node => node.key() === '123_1');
+    const levelBubble2 = wrapper.findWhere(node => node.key() === '456_2');
+    const levelBubble3 = wrapper.findWhere(node => node.key() === '999_3');
+    expect(levelBubble1.find(ProgressTableLevelBubble)).to.have.length(1);
+    expect(levelBubble2.find(ProgressTableLevelBubble)).to.have.length(2); // 1 for level, 1 for sublevel
+    expect(levelBubble3.find(ProgressTableLevelBubble)).to.have.length(1);
+  });
+
+  it('disables progress bubble if there is a bonus and stageExtrasEnabled is false', () => {
+    const wrapper = shallow(<ProgressTableDetailCell {...DEFAULT_PROPS} />);
+    const levelBubble3 = wrapper.findWhere(node => node.key() === '999_3');
+    const progressBubble = levelBubble3.find(ProgressTableLevelBubble);
+    expect(progressBubble.props().disabled).to.equal(true);
+  });
+
+  it('disable is false for progress bubble if there is a bonus and stageExtrasEnabled is true', () => {
+    const props = {...DEFAULT_PROPS, stageExtrasEnabled: true};
+    const wrapper = shallow(<ProgressTableDetailCell {...props} />);
+    const levelBubble3 = wrapper.findWhere(node => node.key() === '999_3');
+    const progressBubble = levelBubble3.find(ProgressTableLevelBubble);
+    expect(progressBubble.props().disabled).to.equal(false);
+  });
+
+  it('generates the right url for level bubble', () => {
+    const wrapper = shallow(<ProgressTableDetailCell {...DEFAULT_PROPS} />);
+    const levelBubble1 = wrapper.findWhere(node => node.key() === '123_1');
+    const url = levelBubble1.find(ProgressTableLevelBubble).props().url;
+    expect(url).to.equal('/level-1?section_id=123&user_id=1');
+  });
+
+  it('calls firehouse putRecord when clicking a level', () => {
+    const wrapper = shallow(<ProgressTableDetailCell {...DEFAULT_PROPS} />);
+    const levelBubble1 = wrapper
+      .findWhere(node => node.key() === '123_1')
+      .childAt(0);
+    levelBubble1.simulate('click');
+    expect(firehoseClient.putRecord).to.have.been.called;
+  });
+
+  it('calls firehouse putRecord when clicking a sublevel', () => {
+    const wrapper = shallow(<ProgressTableDetailCell {...DEFAULT_PROPS} />);
+    const sublevel = wrapper.findWhere(
+      node => node.key() === `${SUBLEVEL_1.id}`
+    );
+    sublevel.simulate('click');
+    expect(firehoseClient.putRecord).to.have.been.called;
+  });
+
+  it('widthForLevels returns the correct width when there are unplugged levels', () => {
+    const levels = [LEVEL_1, LEVEL_1];
+    expect(ProgressTableDetailCell.widthForLevels(levels)).to.equal(250);
+  });
+
+  it('widthForLevels returns the correct width when there are no unplugged levels', () => {
+    const notUnpluggedLevel = {...LEVEL_1, isUnplugged: false};
+    const levels = [notUnpluggedLevel, notUnpluggedLevel];
+    expect(ProgressTableDetailCell.widthForLevels(levels)).to.equal(84);
```

Comment:
if it's not too much work, could you import the various constants that are used to calculate these widths rather than testing with these magic numbers?

Reply:
The only test where this is tricky are for the unplugged levels, the width is calculated by https://github.com/code-dot-org/code-dot-org/blob/cforkish/LP-1604-refactor-progress-tables/apps/src/templates/sectionProgress/progressTables/ProgressTableDetailCell.jsx#L168. Do you think I should export this function and use it in the test?

I updated the other 2.

Reply:
why don't you go ahead and use our `unitTestExports` convention to export that function. however, since there's no constant to compare to, just put a comment saying the 250 is the observed render width for the english localization.

then in a separate test you can choose a language with a noticeably different localization length, and expect that it's wider or narrower than the english localization.

Reply:
I exported the `getUnpluggedWidth` function and tested it independently and then was able to use it to determine the expected width in the unit test.